src/cio_websocket.c

```diff
@@ -0,0 +1,466 @@
+				//TODO: log_err("No start frame was send!");
+				//TODO: handle_error(s, WS_CLOSE_PROTOCOL_ERROR);
+				return;
+			}
+		}
+	}
+
+	if (unlikely(ws->ws_flags.is_fragmented && ((ws->ws_flags.opcode < CIO_WEBSOCKET_CLOSE_FRAME) && (ws->ws_flags.opcode > CIO_WEBSOCKET_CONTINUATION_FRAME)))) {
+		// log_err("Opcode during fragmentation must be 0x0!");
+		// handle_error(s, WS_CLOSE_PROTOCOL_ERROR);
+		return;
+	}
+
+	bool last_frame = false;
+	unsigned int opcode;
+	if (ws->ws_flags.opcode == CIO_WEBSOCKET_CONTINUATION_FRAME) {
+		opcode = ws->ws_flags.frag_opcode;
+		if (unlikely(ws->ws_flags.fin ==1)) {
+			last_frame = true;
+		}
+	} else {
+		opcode = ws->ws_flags.opcode;
+	}
+
+	switch (opcode) {
+	case CIO_WEBSOCKET_BINARY_FRAME:
+		if (likely(ws->onbinaryframe_received != NULL)) {
+			ws->onbinaryframe_received(ws, data, length, last_frame);
+		} else {
+			// TODO: handle_error(s, WS_CLOSE_UNSUPPORTED);
+		}
+
+		break;
+
+	case CIO_WEBSOCKET_TEXT_FRAME:
+		if (likely(ws->on_textframe_received != NULL)) {
+			ws->on_textframe_received(ws, (char *)data, length, last_frame);
+		} else {
+			// TODO: handle_error(s, WS_CLOSE_UNSUPPORTED);
+		}
+
+		break;
+
+	case CIO_WEBSOCKET_PING_FRAME:
+		if (unlikely(length > WS_SMALL_FRAME_SIZE)) {
+			//TODO: handle_error(s, WS_CLOSE_PROTOCOL_ERROR);
+		} else {
+			// TODO: send_pong_frame(ws, data, length);
+			if (ws->on_ping != NULL) {
+				ws->on_ping(ws, data, length);
+			}
+		}
+
+		break;
+
+	case CIO_WEBSOCKET_PONG_FRAME:
+		if (unlikely(length > WS_SMALL_FRAME_SIZE)) {
+			//TODO: handle_error(s, WS_CLOSE_PROTOCOL_ERROR);
+		} else {
+			if (ws->on_pong != NULL) {
+				ws->on_pong(ws, data, length);
+			}
+		}
+
+		break;
+
+	case CIO_WEBSOCKET_CLOSE_FRAME: {
+		uint16_t status_code = CIO_WEBSOCKET_CLOSE_NORMAL;
+		if (length >= 2) {
+			memcpy(&status_code, data, sizeof(status_code));
+			status_code = cio_be16toh(status_code);
+		}
+
+		if (length > 2) {
+			// TODO: struct cjet_utf8_checker c;
+			// TODO: cjet_init_checker(&c);
+			// TODO: if (!cjet_is_byte_sequence_valid(&c, frame + 2, length - 2, true)) {
+			// TODO: 	handle_error(s, WS_CLOSE_UNSUPPORTED_DATA);
+			// TODO: }
+		}
+
+		if ((length == 1) || (length > WS_SMALL_FRAME_SIZE) || is_status_code_invalid(status_code)) {
+			//handle_error(s, WS_CLOSE_PROTOCOL_ERROR);
+		}
+
+		if (ws->self_initiated_close) {
+			ws->close_timer.cancel(&ws->close_timer);
+		} else {
+			if (ws->on_close != NULL) {
+				ws->on_close(ws, (enum cio_websocket_status_code)status_code, NULL);
+			}
+
+			send_close_frame(ws, CIO_WEBSOCKET_CLOSE_GOING_AWAY);
+		}
+
+		break;
+	}
+
+	default:
+		// TODO: handle_error(s, WS_CLOSE_PROTOCOL_ERROR);
+		break;
+
+	}
+}
+
+static void get_payload(struct cio_buffered_stream *bs, void *handler_context, enum cio_error err, struct cio_read_buffer *buffer)
+{
+	(void)bs;
+
+	size_t len = cio_read_buffer_get_transferred_bytes(buffer);
+	if (unlikely((err != CIO_SUCCESS) || (len == 0))) {
+		// TODO: fill out
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this "TODO" comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=c%3AS1135)